### PR TITLE
Issue 120 - Widget

### PR DESCRIPTION
**Pull request by kramer**
#120 - A simple widget with a "Love Track" button and "Now Scrobbling" info. If widget size is smaller than 3 columns then only the button is shown (the number used to determine number of cells -that is 179 in code- is [supposedly not a magic number](http://developer.android.com/guide/practices/ui_guidelines/widget_design.html#anatomy) (70x3-30=180).

Tested only on my Nexus 5.

Please share your opinion.

### CODE REVIEW

**Comment by ilya-murzinov** [comment]
@kramer, please rebase this.

**Comment by kramer** [comment]
@ilya-murzinov done

**Comment by ilya-murzinov** [comment]
@kramer please add all strings translations as well. Just copy-paste them from the main strings.xml.

**Comment by artem-zinnatullin** [comment]
I'll take a look at this tonight.

**Comment by artem-zinnatullin** [comment]
@kramer can you please add some screenshots as comment to the PR?

File: wail-app/src/main/java/com/artemzin/android/wail/service/WAILService.java [hunk 1]
```diff
@@ -109,6 +112,29 @@ public int onStartCommand(Intent intent, int flags, int startId) {
         return START_STICKY;
     }
 
+    private void updateWidget(Track track) {
```
Comment (artem-zinnatullin) [inline wail-app/src/main/java/com/artemzin/android/wail/service/WAILService.java]: Can you please add `@Nullable` annotation to the `track`?

File: wail-app/src/main/java/com/artemzin/android/wail/ui/activity/WAILLoveWidget.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+import android.widget.Toast;
+import com.artemzin.android.wail.R;
+import com.artemzin.android.wail.service.WAILService;
+import com.artemzin.android.wail.storage.WAILSettings;
+import com.artemzin.android.wail.storage.db.LovedTracksDBHelper;
+import com.artemzin.android.wail.storage.model.Track;
+
+public class WAILLoveWidget extends AppWidgetProvider {
+
+    private static final String LOVE_CLICKED = "LOVE_CLICKED";
```
Comment (artem-zinnatullin) [inline wail-app/src/main/java/com/artemzin/android/wail/ui/activity/WAILLoveWidget.java]: Rename to `LOVE_TRACK_BUTTON_CLICKED`?

File: wail-app/src/main/java/com/artemzin/android/wail/ui/activity/WAILLoveWidget.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        }
+    }
+
+    @TargetApi(Build.VERSION_CODES.JELLY_BEAN)
+    @Override
+    public void onAppWidgetOptionsChanged(Context context, AppWidgetManager appWidgetManager, int appWidgetId, Bundle newOptions) {
+        RemoteViews views = new RemoteViews(context.getPackageName(), R.layout.waillove_widget);
+
+        int newMinWidth = newOptions.getInt(AppWidgetManager.OPTION_APPWIDGET_MIN_WIDTH);
+        if (newMinWidth > 179) {
```
Comment (artem-zinnatullin) [inline wail-app/src/main/java/com/artemzin/android/wail/ui/activity/WAILLoveWidget.java]: Please add comment about `179` :)

File: wail-app/src/main/java/com/artemzin/android/wail/ui/activity/WAILLoveWidget.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    public void onReceive(@NonNull Context context, @NonNull Intent intent) {
+        super.onReceive(context, intent);
+
+        if (LOVE_CLICKED.equals(intent.getAction())) {
+            Track track = WAILSettings.getNowScrobblingTrack(context);
+            if (track == null) {
+                Toast.makeText(context, context.getString(R.string.widget_nothing_to_love), Toast.LENGTH_SHORT).show();
+            } else {
+                Toast.makeText(context, context.getString(R.string.main_track_loved), Toast.LENGTH_SHORT).show();
+                LovedTracksDBHelper.getInstance(context).add(track);
```
Comment (artem-zinnatullin) [inline wail-app/src/main/java/com/artemzin/android/wail/ui/activity/WAILLoveWidget.java]: I think work with the DB should be done asynchronously

**Comment by artem-zinnatullin** [comment]
LGTM, but comments!

**Comment by kramer** [comment]
Screenshots:

* 1 cell:
screenshot_2015-08-02-22-35-17

* 4 cells (no info shown when app is disabled or never opened):
screenshot_2015-08-02-22-35-37

* 4 cells with nothing scrobbling info:
screenshot_2015-08-02-22-36-17

* 4 cells with now scrobbling info:
screenshot_2015-08-02-22-35-52

* 3 cells:
screenshot_2015-08-02-22-36-01

* 2 cells:
screenshot_2015-08-02-22-36-08

**Comment by kramer** [comment]
Will apply the changes you mentioned @artem-zinnatullin, thanks for the feedback. I want to emphasise again that this should be tested on a device other than my N5 :).

**Comment by artem-zinnatullin** [comment]
Nice! I'll just add `gravity=center_vertical` for 3rd screenshot

**Comment by kramer** [comment]
> I'll just add gravity=center_vertical for 3rd screenshot

Yup that sticks out :) All the solutions I could think of for that included having some layout related/changing code in `WAILService.updateWidget` and I was not quite comfortable and sure with having that.

**Comment by artem-zinnatullin** [comment]
@kramer may be just display "Now scrobbling" as song name and "Nothing" as artist name?

**Comment by ilya-murzinov** [comment]
@kramer, this is very good job, thank you!

**Comment by kramer** [comment]
@artem-zinnatullin did exactly that now.

@ilya-murzinov You're welcome, glad to contribute.

**Comment by ilya-murzinov** [comment]
I don't have time to take a closer look.
@artem-zinnatullin, please merge this if you think it's ok.

**Comment by artem-zinnatullin** [comment]
@ilya-murzinov unfortunately, I have problems with free time too :(

I think it's okay, but would be nice if we could test it on more devices.

**Comment by artem-zinnatullin** [comment]
I think we can say that widget is in beta mode and ship it :shipit: to the users!

Thanks @kramer!

**Comment by kramer** [comment]
@artem-zinnatullin #120 is still open though...

**Comment by ilya-murzinov** [comment]
@kramer, if you wonder why didn't it close automatically, you should have written "Resolves/closes/solves #120" in first comment for this.
I'll close it.

**Comment by kramer** [comment]
@ilya-murzinov I realised that just a bit late, thanks!

**Pull request merged**